### PR TITLE
Issue #1019: Harden localStorage with schema versioning + QuotaExceededError

### LINKED ISSUE #1019: web: harden localStorage layer — schema versioning + QuotaExceededError
## Background

`webpage_v2/src/services/storage.ts` persists user state across all the keys documented in `webpage_v2/CLAUDE.md`:

```
trackrat:recentTrips
trackrat:favorites
trackrat:favoriteRoutes
trackrat:lastRoute
trackrat:systems
trackrat:homeStation
trackrat:workStation
trackrat:tripHistory
```

Two related weaknesses I confirmed by reading the file:

### 1. `QuotaExceededError` is silently swallowed

Some `localStorage.setItem` calls in `storage.ts` are not wrapped in try/catch at all (they propagate uncaught), and others are caught with bare `catch { /* ignore */ }` blocks (e.g., `appStore.ts:63-68`). On budget devices and constrained browser modes, storage runs out — and the user silently loses their favorites, recent trips, or commute profile with no recovery and no warning.

### 2. No schema version

JSON-deserialized blobs are cast to a TypeScript type with no runtime check. A breaking type change (e.g., `lastUsed: string` → `lastUsed: Date`, or adding a required field) silently constructs invalid objects from old user data and either crashes at first access or, worse, looks fine until much later.

### Versioning

Wrap each get/set with a versioned envelope:

```ts
type Stored<T> = { v: number; data: T };

function readStored<T>(key: string, currentVersion: number, migrate: (raw: unknown) => T): T | null {
  const raw = localStorage.getItem(key);
  if (!raw) return null;
  try {
    const parsed = JSON.parse(raw) as Stored<unknown>;
    if (parsed.v === currentVersion) return parsed.data as T;
    return migrate(parsed); // explicit, per-key migration
  } catch {
    localStorage.removeItem(key);
    return null;
  }
}

function writeStored<T>(key: string, version: number, data: T): void {
  try {
    localStorage.setItem(key, JSON.stringify({ v: version, data }));
  } catch (err) {
    if (err instanceof DOMException && err.name === 'QuotaExceededError') {
      handleQuotaExceeded(key); // see below
    }
    // swallow — non-quota errors here mean storage is unavailable (private mode); accept silent failure
  }
}
```

### Quota handling

Realistic strategy:
- For trip-history-style growing lists, drop the oldest 25% of entries and retry once.
- For singleton blobs (`lastRoute`), nothing to drop — surface a one-time toast/banner via the existing UI (or just accept failure).

### Per-key versions

Each key's reader can have its own version + migrator. Even trivial migrations like "if `lastUsed` is a string, parse it as Date" work for the v1→v2 case.

## Where the change goes

- `webpage_v2/src/services/storage.ts` — central abstraction.
- `webpage_v2/src/store/appStore.ts:63-68` — replace the bare `catch { }` with the new helpers.
- `webpage_v2/src/services/storage.test.ts` already exists — extend with cases:
  - reading old (unversioned) data triggers the migration path
  - QuotaExceededError on `setItem` is handled gracefully (mock `localStorage.setItem` to throw a DOMException of that name)
  - readStored returns `null` and clears the key when JSON.parse fails on corrupted data

## Acceptance

- All eight stored keys go through the versioned envelope.
- Quota errors are caught and either evict-and-retry or fail silently (depending on key).
- A migration path exists (even if currently a no-op) for v1.
- Existing storage tests still pass; new tests cover the migration + quota paths.

## Out of scope

- IndexedDB migration (project explicitly opts out per CLAUDE.md).
- Server-side user accounts.

## Risk note

This touches reader code that runs at app boot. Make sure the migration path is idempotent (`v1` data read after migration writes `v2`, but the read shape doesn't change). Consider testing manually by filling localStorage with the *current* shape and deploying the change locally — the user should not see anything different.

### PR DESCRIPTION
**Pull request by trackrat-dev**
## Summary

- **Schema versioning**: All 8 localStorage keys now use a `{ v: 1, data: T }` envelope. Reads transparently handle legacy (unversioned) data — no user-visible migration needed.
- **QuotaExceededError handling**: All write paths catch `QuotaExceededError`. List-type keys (recentTrips, favoriteRoutes, favorites, tripHistory) evict the oldest 25% of entries and retry once. Singleton keys (lastRoute, systems, homeStation, workStation) fail silently.
- **State ordering fix**: `setHomeStation` and `setWorkStation` in appStore now set in-memory state *before* persisting to storage, matching the pattern used by all other actions.
- **Dead code cleanup**: Removed now-redundant try/catch wrappers in appStore since all storage methods handle errors internally.

## Files Modified

| File | Why |
|------|-----|
| `webpage_v2/src/services/storage.ts` | Core change: `readData`/`writeData` private methods with versioned envelope + quota handling. All public methods refactored to use them. |
| `webpage_v2/src/store/appStore.ts` | Fix ordering bug in `setHomeStation`/`setWorkStation`; remove dead try/catch blocks. |
| `webpage_v2/src/services/storage.test.ts` | Extended from 22 to 42 tests covering versioning, migration, quota errors, preferred systems, limits. |
| `webpage_v2/src/store/appStore.test.ts` | Updated 2 assertions to account for versioned envelope in raw localStorage reads. |

## Test Coverage Added

- **Schema versioning** (8 tests): Legacy unversioned reads for trips/routes/systems/stations, versioned reads, versioned write verification, corrupted data cleanup
- **QuotaExceededError** (5 tests): Eviction + retry for lists, silent failure for singletons, full quota exhaustion, non-quota DOMException
- **Preferred systems** (4 tests): Empty state, save/retrieve, overwrite, corrupted data
- **Trip history limit** (1 test): Verifies 50-entry cap
- **Favorite routes limit** (1 test): Verifies 10-entry cap
- **Trip history dedup** (1 test): Verifies id-based dedup with most-recent-wins

## Design Decisions

- **Version = 1 for all keys**: No per-key versioning needed yet since all data shapes are unchanged. The migration path exists for future schema changes.
- **Evict oldest 25%**: Pragmatic balance — drops enough to likely succeed on retry without losing most user data.
- **No toast/banner on quota**: Silent failure matches the existing pattern for this app (no backend user accounts, localStorage is best-effort).

Closes #1019

https://claude.ai/code/session_01HZAHX2SnYEUbyqpGj9F5UB

---
_Generated by [Claude Code](https://claude.ai/code/session_01HZAHX2SnYEUbyqpGj9F5UB)_

### CODE REVIEW

File: webpage_v2/src/services/storage.ts [hunk 1]
```diff
@@ -14,26 +14,54 @@ const MAX_RECENT_TRIPS = 10;
...
+        } catch { /* retry failed — storage fully unavailable */ }
+      }
     }
   }
 
+  // Recent Trips
+  getRecentTrips(): TripPair[] {
+    const trips = this.readData<TripPair[]>(RECENT_TRIPS_KEY);
+    if (!trips) return [];
+    return trips.map((trip) => ({
```
[the remark this answers is not in the document]
Comment (trackrat-dev) [reply webpage_v2/src/services/storage.ts:56]: Valid concern. `readData` returns `parsed as T` without type checking, so if localStorage contains valid-but-wrong JSON (e.g. `{}` or `0`), all list getters would throw `TypeError` on `.map()`.

Fixed in `11edd68`: replaced `if (!trips) return []` with `if (!Array.isArray(trips)) return []` in all five list getters (`getRecentTrips`, `getFavoriteRoutes`, `getFavoriteStations`, `getPreferredSystems`, `getTripHistory`). `Array.isArray` covers both the `null` case (from `readData` returning null) and the non-array-JSON case.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: webpage_v2/src/services/storage.ts [hunk 2]
```diff
@@ -14,26 +14,54 @@ const MAX_RECENT_TRIPS = 10;
...
+      if (!raw) return null;
+      const parsed = JSON.parse(raw);
+      if (parsed && typeof parsed === 'object' && typeof parsed.v === 'number' && 'data' in parsed) {
+        return parsed.data as T;
+      }
+      return parsed as T;
     } catch {
-      return [];
+      localStorage.removeItem(key);
+      return null;
```
[the remark this answers is not in the document]
Comment (trackrat-dev) [reply webpage_v2/src/services/storage.ts:35]: Valid concern. If `localStorage.getItem()` throws a `SecurityError` (restricted browser, cross-origin iframe, storage disabled), the catch block's `localStorage.removeItem(key)` would throw the same exception — escaping the handler and breaking the safe-fallback contract.

Fixed in `11edd68`: wrapped `removeItem` in its own try/catch:
```typescript
catch {
  try { localStorage.removeItem(key); } catch { /* storage inaccessible */ }
  return null;
}
```

Callers now always get `null` regardless of how storage fails.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: webpage_v2/src/services/storage.ts [hunk 3]
```diff
@@ -88,111 +106,95 @@ class StorageService {
...
+    if (favorites.some(f => f.id === station.id)) return;
 
     const newFavorite: FavoriteStation = {
       ...station,
       addedDate: new Date(),
     };
 
     favorites.push(newFavorite);
-    localStorage.setItem(FAVORITES_KEY, JSON.stringify(favorites));
+    this.writeData(FAVORITES_KEY, favorites, evictOldest);
```
[the remark this answers is not in the document]
Comment (trackrat-dev) [reply webpage_v2/src/services/storage.ts]: Valid concern. `evictOldest` keeps the first 75% via `slice(0, ...)`, and `addFavoriteStation` was using `push` (appending to tail) — so the just-added favorite would be the first to get evicted. The other callers (`saveRecentTrip`, `saveFavoriteRoute`, `saveTripHistoryEntry`) all prepend, so they were fine.

Fixed in `11edd68`: changed `favorites.push(newFavorite)` to `favorites.unshift(newFavorite)`, matching the pattern used by all other callers. Now `evictOldest` correctly drops the oldest favorites (at the tail) and retains the newest (at the head, including the just-added one).

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by trackrat-dev** [comment]
## Review feedback summary

Codex raised 3 concerns on this PR — all valid and now addressed in `11edd68`:

| # | Concern | Severity | Fix |
|---|---------|----------|-----|
| 1 | `readData` returns any parseable JSON; list getters call `.map()` without verifying the payload is an array | P1 | Added `Array.isArray()` guards in all 5 list getters |
| 2 | `readData` catch block calls `localStorage.removeItem()` which can throw the same `SecurityError` that triggered the catch | P1 | Wrapped `removeItem` in its own try/catch |
| 3 | `addFavoriteStation` uses `push` (tail) but `evictOldest` keeps the first 75% — newest item gets evicted first | P2 | Changed to `unshift` (head), matching all other callers |

All 203 tests pass after the fixes. **Recommend moving forward with merge** — the original refactoring is solid and the three edge cases are now handled.

---
_Generated by [Claude Code](https://claude.ai/code)_
